[ChanXiaoYong] iP

caxewhy opened the pull request:
# 📝Solyu  

> "Efficiency is doing things right; effectiveness is doing the right things." – [Peter Drucker](https://intranet.engineering.ucdavis.edu/sites/g/files/dgvnsk9601/files/inline-files/Effectiveness-vs.-efficiency-%E2%80%93-Let%E2%80%99s-not-confuse-the-two-1.pdf)  

Solyu is a **task management application** optimized for fast typists. It provides a simple yet powerful **Command Line Interface (CLI)** with GUI feedback.

---

## 🚀 Features  

- [x] **Task Management**: Create and track tasks.  
- [x] **Different Task Types**: Supports ToDos, Deadlines, and Events.  
- [x] **Task Persistence**: Saves tasks locally.  
- [x] **Find Feature**: Quickly locate tasks by keywords.  
- [x] **User-friendly CLI**: Command-based interactions for efficiency.  
- [ ] **Upcoming**: ~~Task reminders~~  

### All you need to do is:

1.  **Download it from** [here](https://github.com/nus-cs2103-AY2425S2/ip/releases)  
2.  **Double-click** to open  
3.  **Add your tasks**  
4.  **Let it manage your tasks for you** 😃  

And best of all... **it's FREE!** 🎉 

## 🛠 Setup Instructions  

- **Download** the latest JAR file from [here](https://github.com/nus-cs2103-AY2425S2/ip/releases).  
- **Place** the JAR file in your chosen folder.  
- **Open** a terminal and navigate to the folder.  
- **Run** the following command:  
   ```sh
   java -jar solyu.jar


If you are a Java programmer, you can use this project to practice **Java programming**. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Did Level 0. Rename, Greet, Exit.

Commit:
Did Level 1. Added Echo commands.

Commit:
Did Level 2. Ability to add tasks, and List them.

Commit:
Did Level 3. Added ability to mark/unmark tasks as done.

Commit:
Did Level 4. Added ability to create ToDos, Events, Deadlines.

Commit:
Did Automated Text UI Testing.

Commit:
Did Level 5. Taught chatbot how to handle errors such as incorrect inputs entered by the user.

Commit:
Did Level 6. Ability to delete tasks from list.

Commit:
Added ability to save and load tasks from text file.

Commit:
Merge branch 'branch-Level-7'

Commit:
Did Level 8. Dates and Times.

Commit:
Merge branch 'branch-Level-8'

Commit:
A-MoreOOP. Made code more OOP.

Commit:
Merge branch 'add-gradle-support'

Commit:
Add JUnit tests.

Commit:
Package the App as a JAR file.

Commit:
Add JavaDoc comments to the code.

Commit:
Add JavaDoc comments to the code.

Commit:
Give users a way to find a task by searching for a keyword in the task description.

Commit:
Merge branch 'branch-Level-9'

Commit:
Tweak the code to comply with a coding standard

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/solyu/Deadline.java
#	src/main/java/solyu/Solyu.java

Commit:
Add checkstyle plugin

Commit:
Merge branch 'A-Checkstyle'

Commit:
Add a GUI to the chatbot

Commit:
Merge branch 'branch-Level-10'

JunXiang26 commented on src/main/java/solyu/Solyu.java:
```diff
@@ -0,0 +1,282 @@
+        String command = parsedInput[0];
+        String argument = parsedInput[1];
+
+        switch (command) {
+        case "greet":
+            return ui.getGreetingMessage();
+
+        case "find":
+            return handleFind(argument);
+
+        case "add":
+            return handleTaskAdd(argument, new Task(argument));
+
+        case "todo":
+            return handleTodo(argument);
+
+        case "deadline":
+            return handleDeadline(argument);
+
+        case "event":
+            return handleEvent(argument);
+
+        case "list":
+            return ui.getListAsString(taskList.getTasks());
+
+        case "mark":
+            return handleTaskMark(argument, true);
+
+        case "unmark":
+            return handleTaskMark(argument, false);
+
+        case "delete":
+            return handleTaskDelete(argument);
+
+        case "bye":
+            storage.saveTasksToFile(new ArrayList<>(taskList.getTasks()));
+            return ui.getGoodbyeMessage();
+
+        default:
+            return ui.getErrorMessage(ERROR_UNKNOWN_COMMAND + command + "\nPlease try another command :)");
```
Nice work on indenting the cases to the switch statement!

JunXiang26 commented on src/main/java/solyu/Deadline.java:
```diff
@@ -0,0 +1,78 @@
+package solyu;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
```
Nice work importing seperate libraries!

JunXiang26 commented on src/main/java/solyu/DialogBox.java:
```diff
@@ -0,0 +1,85 @@
+package solyu;
+
+import javafx.geometry.Pos;
+import javafx.scene.control.Label;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
```
Same thing with the imports being seperated! Great work!

JunXiang26 commented on src/main/java/solyu/Deadline.java:
```diff
@@ -0,0 +1,78 @@
+     * @return The parsed LocalDate object.
+     * @throws IllegalArgumentException if the date format is invalid.
+     */
+    private LocalDate parseDateTime(String dateStr) {
+        try {
+            DateTimeFormatter format = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+            return LocalDate.parse(dateStr, format);
+        } catch (DateTimeParseException e) {
+            throw new IllegalArgumentException(
+                    "Error: Invalid date format! Please use yyyy-MM-dd (e.g., 2024-02-05)."
+            );
+        }
+    }
```
Nice work using Egyptian style brackets!

JunXiang26 commented on src/main/java/solyu/Solyu.java:
```diff
@@ -0,0 +1,282 @@
+ */
+public class Solyu {
+    // Constants for error messages
+    private static final String ERROR_EMPTY_COMMAND =
```
Great work with Screaming Variable Naming for final variables.

JunXiang26 reviewed (commented):
Overall a great work on coding standards! Nice job!

Commit:
Add GUI elements

Commit:
Add assertions to verify critical logic assumptions

Assertions have been added to key methods in Solyu to ensure
that important assumptions hold during execution. These include:

- Verifying that task indices are within bounds before accessing tasks.
- Ensuring non-null values for critical objects.
- Checking that command arguments are not empty before processing.

This helps catch logical errors during development and debugging
when assertions are enabled (`-ea`). These assertions do not
replace proper error handling but serve as internal consistency checks.

With this change, the program will now halt early if an assertion
fails, making it easier to detect bugs before they affect execution.

Commit:
Improve code quality by refactoring code

Several sections of the code have been improved to enhance readability, maintainability, and adherence to best practices. Key changes include:

- Applied meaningful constant names: Replaced magic strings with descriptive constants for error messages and UI labels.
- Improved variable naming: Renamed vague variables for clarity, making the code self-documenting.

These refinements make the codebase easier to read, maintain, and extend in future iterations.

Commit:
Refactor code to use Java Streams

Previously, various parts of the code relied on traditional loops and manual list manipulations, making the logic verbose and harder to maintain. This commit introduces Java Streams to improve code clarity and efficiency.

Key changes:
- Updated `Storage#saveTasksToFile` to use `Files.write()` with `stream().map().toList()`, replacing `PrintWriter`.
- Modified `TaskList#markTask` to use `stream().skip(index).findFirst().ifPresent(Task::markAsDone)`, reducing direct list indexing.

These changes enhance code maintainability by reducing boilerplate loops, improving readability, and leveraging functional programming best practices. The behavior remains unchanged while ensuring better future scalability.

caxewhy committed:
Merge pull request #1 from caxewhy/branch-A-Assertions

Add assertions to verify critical logic assumptions

Commit:
Merge branch 'master' of https://github.com/caxewhy/ip

Commit:
Merge branch 'branch-A-Streams'

Commit:
Merge branch 'branch-A-CodeQuality'

Commit:
Add detect duplicates extension

Commit:
Improve the GUI to make it more polished

Commit:
Give a unique personality to Solyu

Commit:
Merge branch 'branch-A-Personality'

Commit:
Add error handling code

Commit:
Add a User Guide

Commit:
Amend regex characters

caxewhy committed:
Update README.md

caxewhy committed:
Update README.md

Commit:
Update readme

Commit:
Merge branch 'master' of https://github.com/caxewhy/ip

Commit:
Update launcher

Commit:
Fix minor storage issues

Commit:
Remove unuse imports

Commit:
Fix index out of range issue

Commit:
Update user guide

Commit:
Fix user guide
